src/fingerprint.ts

```diff
@@ -0,0 +1,223 @@
+
+import type {
+  ScriptConfig,
+  ScriptReference,
+  ScriptReferenceString,
+} from './script.js';
+
+/**
+ * All meaningful inputs of a script. Used for determining if a script is fresh,
+ * and as the key for storing cached output.
+ */
+export interface FingerprintData {
+  /**
+   * Brand to make it slightly harder to create one of these interfaces, because
+   * that should only ever be done via {@link Fingerprint.compute}.
+   */
+  __FingerprintDataBrand__: never;
+
+  /**
+   * Whether the output for this script can be fresh or cached.
+   *
+   * True only if the "files" array was defined for this script, and for all of
+   * this script's transitive dependencies.
+   */
+  cacheable: boolean;
+
+  /** E.g. linux, win32 */
+  platform: NodeJS.Platform;
+
+  /** E.g. x64 */
+  arch: string;
+
+  /** E.g. 16.7.0 */
+  nodeVersion: string;
+
+  /**
+   * The shell command from the Wireit config.
+   */
+  command: string | undefined;
+
+  /**
+   * The "clean" setting from the Wireit config.
+   *
+   * This is included in the fingerprint because switching from "false" to "true"
+   * could produce different output, so a re-run should be triggered even if
+   * nothing else changed.
+   */
+  clean: boolean | 'if-file-deleted';
+
+  // Must be sorted.
+  files: {[packageDirRelativeFilename: string]: Sha256HexDigest};
+
+  /**
+   * The "output" glob patterns from the Wireit config.
+   *
+   * This is included in the fingerprint because changing the output patterns
+   * could produce different output when "clean" is true, and because it affects
+   * which files get included in a cache entry.
+   *
+   * Note the undefined vs empty-array distinction is not meaningful here,
+   * because both cases cause no files to be deleted, and the undefined case is
+   * never cached anyway.
+   */
+  output: string[];
+
+  // Must be sorted.
+  dependencies: {[dependency: ScriptReferenceString]: FingerprintData};
+}
+
+/**
+ * String serialization of a {@link FingerprintData}.
+ */
+export type FingerprintString = string & {
+  __FingerprintStringBrand__: never;
+};
+
+/**
+ * SHA256 hash hexadecimal digest of a file's content.
+ */
+export type Sha256HexDigest = string & {
+  __Sha256HexDigestBrand__: never;
+};
+
+/**
+ * The fingerprint of a script. Converts lazily between string and data object
+ * forms.
+ */
+export class Fingerprint {
+  static fromString(string: FingerprintString): Fingerprint {
+    const fingerprint = new Fingerprint();
+    fingerprint.#str = string;
+    return fingerprint;
+  }
+
+  /**
+   * Generate the fingerprint data object for a script based on its current
+   * configuration, input files, and the fingerprints of its dependencies.
+   */
+  static async compute(
+    script: ScriptConfig,
+    dependencyFingerprints: Array<[ScriptReference, Fingerprint]>
+  ): Promise<Fingerprint> {
+    let allDependenciesAreCacheable = true;
+    const filteredDependencyStates: Array<
```

Comment:
Should this be renamed to `filteredDependencyFingerprints`?

Reply:
Done thanks.